---
layout: tutorial_hands_on
topic_name: Introduction to RNAseq
tutorial_name: intro_rnaseq
---

# Introduction

## RNAseq definition

The type of information you can get from RNAseq can be of broadly two type: Qualitative and quantitative data : 
*   Qualitative: Includes identification of expressed transcripts or intro/exon boundaries, transcriptional start sites (TSS), and poly-A sites... Can also be referred to as annotation. The most important parameter in these analyses is that the reads cover evenly the transcripts. 
*   Quantitative : Includes measurement of difference of expression, alternative splicing, alternative TSS, and alternative polyadenylation between two or more treatments or groups... 


> ### :pencil2: Typical Analysis pipeline
>
> 1. Initial processing : Demultiplex, filter, and trim sequencing reads.
> 2. Normalize sequencing reads (if performing de novo assembly)
> 3. de novo assembly of transcripts (if a reference genome is not available) or Map (align) sequencing reads to reference genome or transcriptome
> 4. Annotate transcripts assembled or to which reads have been mapped
> 5. Count mapped reads to estimate transcript abundance
> 6. Perform statistical analysis to identify differential expression (or differential splicing) among samples or treatments
> 7. Perform multivariate statistical analysis/visualization to assess transcriptome-wide differences among samples. 
>
{: .hands_on}

### Initial Processing  

The reads must go through a number of step before de novo assembly or mapping on a reference.

The steps are the following : 
*   Demultiplex by index or barcode
*   Remove adapter sequences
*   Trim reads by quality
*   Discard reads by quality/ambiguity
*   Filter reads by k-mer coverage
*   Normalize k-mer coverage

#### Demultiplexing

Multiplexing is the pooling of several different samples into a single pool of molecules read by the sequencer

"Multiplexing of samples is made possible by incorporation of a short (usually at least 6 nt) index or "barcode"; into each DNA fragment during the adapter ligation or PCR amplification steps of library preparation. After sequencing, each read may be traced back to its original sample using the index sequence and binned accordingly. Depending on the library preparation method and sequencing technology, the sequencing instrument software may or may not perform the partitioning. "

Tools : Kraken, deindexer, Drop-seq, Bayexer, Sabre

In galaxy : Kraken, 

#### Removing adapter

"If cDNA insert sizes are sufficiently small and sequencing read lengths sufficiently long, it is possible to generate sequencing reads that contain a portion of adapter sequence at the 3â€²-end. Therefore, depending on the library preparation method and range of cDNA insert sizes, it may be prudent to trim these artifacts away. Removing these artifacts should improve assembly and/or mapping of sequencing reads. "

Tools : cutadapt, The FASTX-toolkit, Scythe, FAR, BBDuk, clean_reads, condetri, Deconseq, bowtie, htSeqTools, Sabre

In Galaxy : cutadapt, fastx, Scythe, bowtie, 

#### Filtering/trimming

Quality control : AfterQC, FastQC, fastqp, Kraken, HTSeq, mRIN, MultiQC, NGS QC Toolkit, PRINSEQ, RNA-SeQC, Trim galore, 

Tools filtering/trimmimg : AfterQC, NGS QC Toolkit, Sickle, Trimmomatic

In Galaxy : FastQC, Kraken, HTSeq, MultiQC, bamqc, PRINSEQ,  Trim galore, Trimmomatic

#### Error Correction

Tools : AmpliconNoise, BayesHammer, QuorUM, GeneScissors, 

In Galaxy : 

### Assembly 

#### RNAseq *De novo* Assembly

"If you can't align RNA-seq data to a well-assembled reference genome from a relatively recently diverged organism, building a reference transcriptome out of RNA-seq reads is an alternative strategy. The primary goal in assembling a transcriptome de novo is to reconstruct a set of contiguous sequences (contigs) presumed to reflect accurately a large portion of the RNAs actually transcribed in the cells of your organism. [...] The assembly process must place reads from the same transcript together in the face of variants introduced by polymorphism and sequencing errors, and the process must assemble reads from different but often similar, paralogous transcripts as separate contigs"

Tools : 
*    de Bruijn Graph-Based Assembly : Velvet, Oases, Trans-ABySS, Trinity, 
*    Overlap-layout-consensus assembly : CAP3, MIRA 3, Roche GS De Novo Assember ("Newbler")

Trinity
In Galaxy : Velvet, Oases, Trinity, CAP3, MIRA, Newbler

#### Mapping to a reference

"Short reads generated by RNA-seq experiments must ultimately be aligned, or "mapped" to a reference genome or transcriptome assembly. Read alignment to a reference provides biological information in two basic ways. First, it generates a dictionary of the genomic features represented in each RNA-seq sample. That is, aligned reads become annotated, and the highly fragmented data is thus connected to the gene families, individual transcripts, small RNAs, or individual exons encompassed by the original tissue sample. Second, the number of reads aligned to each feature approximates abundances of those features in the original sample. Such measures of digital gene expression are then subject to comparison among samples or treatments in a statistical framework."

Tools :  Bowtie2, BWA-mem, GSnap
In Galaxy : Bowtie2, BWA-mem, GSnap (2011), trinity

##### Mapping to genome

"If you are mapping your reads to a reference genome, you need an alignment tool that can allow for a read to be "split" between distant regions of the reference in the event that the read spans two exons. "

##### Mapping to transcriptome

"If you are mapping reads to a transcriptome intron/exon boundaries are irrelevant. However you may still need to worry about gapped alignments depending on the type of transcriptome you assembled. There are two basic strategies to transcriptome assembly: 1) assemble a single, longest gene model for each set of splice variants, and 2) try to reconstruct each variant separately. "

> ### :bulb: Tip: microRNA alignement
>
>"Some aligners are designed to map especially short sequences, such as those obtained from mircoRNA-seq studies. These aligners include parameter value ranges necessary for the optimal alignment of reads that are only 16-30 nt. "
>
>tools : MicroRazerS, mrFAST, mrsFAST, PatMaN
{: .tip}

In Galaxy :

### Annotation

"Most people are interested in the identities of the transcripts they assemble or identify as differentially expressed or alternatively spliced. Annotations connect novel RNA-seq data with biological information about the same (or homolgous) transcripts derived from other studies. For example, it is important to identify gene families or known pathways that respond transcriptionally to specific experimental treatments. Also, RNA-seq data derived from eukaryotic organisms may contain reads from other species (parasites, for example), and one would want to know about this.

In the event that a reference genome is available, annotation is relatively straightforward. Genomic coordinates from the reference genome are normally associated with various forms of annotation information through databases such as genome browsers. Given mapping coordinates from an alignment, for example, it is possible to extract the corresponding information from such databases."

Tools : Blast2GO, DAVID, ermineJ

### Differential expression

"The ultimate objective of many RNA-seq studies is to identify features of the transcriptome that are differentially expressed between or among groups of individuals or tissues that are different with respect to some condition. It is important to keep in mind the nature of those "features" to which reads were mapped for counting."

Tools : ABSSeq, ALDEx2, Alexa-Seq, BaySeq, Cufflinks/Cuffdiff, DiffSplice
In Galaxy : Blast2GO, Cufflinks/Cuffdiff, 

> ### :bulb: Tip: Normalization of read count
>
>"If comparison of expression values among different transcripts is of interest, other normalization factors must be considered. Gene length, for example, influences read counts because more reads per single transcript will be observed for longer genes. The normalization procedure known as "RPKM" (reads per kilobase per one million mapped reads) applies a gene length and library size adjustment (Mortazavi et al. 2008). Although not formally treated in any differential expression software of which we are aware, base content has also been identified as an important among-gene normalization consideration (Oshlack, Robinson, Young 2010). For a detailed, more comprehensive guide to normalization strategies for RNA-seq data, see Dillies et al. 2012 and Bullard et al. 2010."
>
>tools : 
{: .tip}

#### Discrete distribution model 

"To estimate transcript abundance of a given target in a given sample, a parameter of interest is the probability that a randomly drawn read (from millions in the library) maps uniquely to that target. If thousands of reads in the library map to the particular target, this probability will be higher than if only a few map, but in general these probabilities will be very low owing to the huge size of the library and the complexity of the transcriptome. These sampling properties are characterized by the Poisson probability distribution, which in principle should enable us to draw inferences about, say, whether the probability that a randomly drawn read from a library of type A maps to transcript X is higher than the probability that a randomly drawn read from library type B maps to transcript X (i.e. differential expression of transcript X between groups A and B).

One issue with RNA-seq data, however, is that the variance of this probability among different individuals of a group is substantially higher than the mean, with respect to many genes (Anders, Huber 2010). A Poisson distribution assumes an equal mean and variance, and is therefore not a good fit. This issue, known as "overdispersion," has inspired statistical software authors to adopt other models, particularly the negative binomial (NB) distribution, which is characterized by an additional dispersion parameter. Several popular differential expression packages, such as edgeR (Robinson, McCarthy, Smyth 2010) and DESeq (Anders, Huber 2010) are based on the NB distribution, but they differ extensively in how the dispersion parameter is estimated, how normalization is performed, or how the hypothesis test is carried out. For a nice tutorial on how DESeq works in these respects, and its actual usage, see: cgrlucb.wikispaces.com/Spring+2012+DESeq+Tutorial."

Tools : edgeR, DESeq
In Galaxy  : edgeR, DESeq

#### Continuous distribution model

"Rather than modeling expression estimates as counts, some statistical approaches treat normalized or transformed count values as continuously distributed variables. If these are distributed approximately normally or log-normally, then those distributions may in principle be used for differential expression inferences. Some authors, for example, have compared normalized count data between two groups via t-test (Busby et al. 2013). Because microarray data is continuously distributed, some have used software originally written for array data (Smyth 2004) to analyze normalized RNA-seq data (Soneson, Delorenzi 2013). One upshot of this approach is the ability to use a general linear model framework for more complex experimental designs. For data sets with larger sample sizes (e.g. 5-10), Soneson et al. 2013 found this method to perform especially well."


### Allelic Expression

"quantifies expression variation between the two haplotypes of a diploid individual distinguished by heterozygous sites"

Tools : GATK
In Galaxy : GATK

### SNV Detection

"The only way to be absolutely sure of the individual's mutations is to compare the transcriptome sequences to the germline DNA sequence. This enables the distinction of homozygous genes versus skewed expression of one of the alleles and it can also provide information about genes that were not expressed in the transcriptomic experiment. An R-based statistical package known as CummeRbund[61] can be used to generate expression comparison charts for visual analysis."

Tools : samtoolsn, annovar, GATK
In Galaxy : samtools, annovar, GATK

### Fusion Gene Detection

Tools : 


| Tool Name      | Group                      | Reference                                | 
| -------------- |:--------------------------:|-----------------------------------------:|
| Bellerophontes | Paired-end + Fragmentation | Bioinformatics. (Abate et al. 2012)      |
| BreakFusion    | Whole paired-end           | Bioinformatics. (Chen et al. 2012)       |
| ChimeraScan    | Paired-end + Fragmentation | Bioinformatics. (Iyer et al. 2011)       |
| EricScript     | Whole paired-end           | Bioinformatics. (Benelli et al. 2012)    | 
| FusionCatcher  | Paired-end + Fragmentation | bioRxiv. (Nicorici et al 2012)           | 
| FusionHunter   | Whole paired-end           | Bioinformatics. (Li et al. 2011)         | 
| FusionMap      | Direct Fragmentation       | Bioinformatics. (Ge et al. 2011)         | 
| JAFFA          | Paired-end + single-end    | Genome Medicine. (Davidson et al. 2015)  | 
| MapSplice      | Direct Fragmentation       | Genome Medicine. (Davidson et al. 2010)  | 
| nFuse          | Whole paired-end           | Genome research. (McPherson et al. 2012) | 
| SOAPFuse       | Whole paired-end           | Genome biology. (Jia et al. 2013)        | 
| TopHat-Fusion  | Paired-end + Fragmentation | Genome biology. (Kim and Salzberg, 2011) | 

Table from [Kumar et al. 2016](10.1038/srep21597)

In Galaxy : TopHat-Fusion (post step)


## RNAseq tools

| Software      | De-multiplexing | Adaptor Trimming | Quality Filtering/ Trimming | K-mer Filtering | K-mer Normalization |
| ------------- |:-------------:|:-------------:|:-------------:|:-------------:| -----:| 
| FASTX-Toolkit | [x] | [x] | [x] |     |     |
| Goby          | [x] | [x] |     |     |     |
| khmer         |     |     |     | [x] | [x] |
| NGS_backbone  |     | [x] | [x] |     |     |
| Stacks        | [x] | [x] | [x] | [x] | [x] |
| trimmomatic   |     | [x] | [x] |     |     |
| biopieces     | [x] | [x] | [x] |     |     |

In Galaxy :  fastx, khmer, Stacks, trimmomatic 

Differential expression tools

| Package       | Reference                      | Model                | 
| ------------- |:------------------------------:|---------------------:|
| PoissonSeq    | Li et al. 2012                 | Poisson log-linear   |
| iFad          | Chung et al. 2013              | Poisson & Gamma      |
| TSPM          | Auer, Doerge 2011              | Quasi-Poisson        |
| baySeq        | Hardcastle et al 2010          | NB                   | 
| DESeq2        | Love, Huber, Anders 2014       | NB                   | 
| EBSeq         | Leng et al. 2013               | NB                   | 
| edgeR         | Robinson, McCarthy, Smyth 2010 | NB                   | 
| NBPSeq        | Di et al. 2011                 | NB                   | 
| ShrinkSeq     | Van De Wiel et al. 2012        | NB                   | 
| sSeq          | Yu, Huber, Vitek 2013          | NB                   | 
| BBSeq         | Zhou, Xia, Wright 2011         | NB                   | 
| CuffDiff 2    | Trapnell et al. 2013           | NB                   | 
| QuasiSeq      | Lund et al. 2012               | NB                   | 
| DEGseq        | Wang et al. 2010               | MA-plot (normal)     | 
| LIMMA         | Smyth 2004                     | general linear model | 
| NOISeq        | Tarazona et al. 2011           | nonparametric        | 
| SAMseq        | Li, Tibshirani 2011            | nonparametric        | 

In Galaxy : DESeq2, EBSeq, edgeR, CuffDiff, 

## Let's try it

> ### Agenda
>
> In this tutorial, we will deal with:
>
{: .agenda}


# Sources 

https://rnaseq.uoregon.edu/
https://en.wikipedia.org/wiki/List_of_RNA-Seq_bioinformatics_tools#Quality_control.2C_trimming.2C_error_correction_and_pre-processing_of_data
